### PR TITLE
[US-3633] entities by factories

### PR DESCRIPTION
**Pull request by simara-svatopluk**
| Q             | A
| ------------- | ---
|Description, reason for the PR| US-3633
|New feature| Yes
|BC breaks| No
|Fixes issues| ...
|Standards and tests pass| Yes
|Have you read and signed our [License Agreement for contributions](https://www.shopsys.com/license-agreement)?| Yes

### CODE REVIEW

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+     * @param string $slug
+     * @return \Shopsys\FrameworkBundle\Component\Router\FriendlyUrl\FriendlyUrl
+     */
+    public function create(
+        string $routeName,
+        int $entityId,
+        int $domainId,
+        string $slug
+    ): FriendlyUrl {
+        return new FriendlyUrl($routeName, $entityId, $domainId, $slug);
```
Comment (vitek-rostislav) [inline packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php]: I noticed that there are few entities that do not have corresponding DataObject. Could you please create a new user-story with a list of such entities? I think we should unify the data-object approach or at least discuss it sometime later

File: packages/framework/src/Component/Image/Image.php [hunk 2]
```diff
@@ -77,7 +77,7 @@ class Image implements EntityFileUploadInterface
      * @param string $entityName
      * @param int $entityId
      * @param string|null $type
-     * @param string $temporaryFilename
+     * @param string|null $temporaryFilename
```
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Component/Image/Image.php]: I don't know, but typehint of `$temporaryFilename` is nullable and also method `getTemporaryFilesForUpload` considers `$temporaryFilename` nullability

File: packages/framework/src/Model/Administrator/Activity/AdministratorActivityFacade.php [hunk 3]
```diff
@@ -17,12 +17,25 @@ class AdministratorActivityFacade
      */
     protected $administratorActivityRepository;
 
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Administrator\Activity\AdministratorActivityFactoryInterface
+     */
+    protected $administratorActivityFactory;
+
+    /**
+     * AdministratorActivityFacade constructor.
```
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Model/Administrator/Activity/AdministratorActivityFacade.php]: That's bad, I spent a lot of time to fix constructors :-( Ok, I'll remove at least this line

File: packages/framework/src/Model/Product/Brand/BrandFacade.php [hunk 4]
```diff
@@ -34,18 +34,33 @@ class BrandFacade
      */
     protected $domain;
 
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Product\Brand\BrandFactoryInterface
+     */
+    protected $brandFactory;
+
+    /**
+     * @param \Doctrine\ORM\EntityManagerInterface $em
```
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply packages/framework/src/Model/Product/Brand/BrandFacade.php]: But I'll keep them

File: packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php (hunk 1 above)
Comment (simara-svatopluk) [reply packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php]: This problem is trickier, for example `AdministratorActivity` expects `Administrator $administrator, string $ipAddress`.
These information could be eventually wrapped in data object, but should it be?

Comment (simara-svatopluk) [reply packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php]: This problem is trickier, for example `AdministratorActivity` expects `__construct(Administrator $administrator, string $ipAddress)`.
This also seem to be problem of all `*Domain` objects...
These information could be eventually wrapped in data object, but should it be?

Comment (vitek-rostislav) [reply packages/framework/src/Component/Router/FriendlyUrl/FriendlyUrlFactory.php]: As we discussed earlier, I'll create a US for that, or maybe I'll just add a note into some existing one :wink:

File: docs/introduction/basics-about-model-architecture.md [hunk 5]
```diff
@@ -67,6 +67,44 @@ class Product
 }
 ```
 
+### Factory
```
[the remark this answers is not in the document]
Comment (simara-svatopluk) [reply docs/introduction/basics-about-model-architecture.md]: I'm not sure. It is in section ##Entity, and almost everything in this article is about entities

File: packages/framework/src/Model/Pricing/Currency/CurrencyFacade.php [hunk 6]
```diff
@@ -63,6 +63,10 @@ class CurrencyFacade
      * @var \Shopsys\FrameworkBundle\Model\Payment\PaymentPriceFactoryInterface
      */
     protected $paymentPriceFactory;
+    /**
```
Comment (vitek-rostislav) [inline packages/framework/src/Model/Pricing/Currency/CurrencyFacade.php]: missing blank line :laughing: #nitpick

File: docs/introduction/basics-about-model-architecture.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply docs/introduction/basics-about-model-architecture.md]: yes, you are right :+1:

File: docs/introduction/basics-about-model-architecture.md [hunk 7]
```diff
@@ -233,8 +271,13 @@ class CartService
     /**
      * @var \Shopsys\FrameworkBundle\Model\Product\Pricing\ProductPriceCalculationForUser
      */
-    private $productPriceCalculation;
+    protected $productPriceCalculation;
```
[the remark this answers is not in the document]
Comment (vitek-rostislav) [reply docs/introduction/basics-about-model-architecture.md]: thank you

**Pull request merged**